README.md
<p align="center">
<img src="https://raw.githubusercontent.com/adrianflutur/flutter_improved_scrolling/main/doc/media/logo.png" width="640" alt="improved_scrolling" />
</p>

[![pub package](https://shields.io/pub/v/flutter_improved_scrolling.svg?style=flat-square&color=blue)](https://pub.dev/packages/flutter_improved_scrolling)

### An attempt to implement better scrolling for Flutter Web and Desktop.

### Includes keyboard, MButton and custom mouse wheel scrolling.

<br>

## _**Getting started**_

- [Example](#example)
- [Usage and features](#usage-and-features)
- [License](#license)

<br>

## _**Example**_

<br>

<img src="https://raw.githubusercontent.com/adrianflutur/flutter_improved_scrolling/main/doc/media/improved_scrolling_example.gif" width="1280"/>

<br><br>

## _**Usage and features**_

(from the [example app](https://github.com/adrianflutur/flutter_improved_scrolling/tree/main/example/lib/scrollable_page.dart))

```dart
final controller = ScrollController();

...

ImprovedScrolling(
      scrollController: controller,
      onScroll: (scrollOffset) => debugPrint(
        'Scroll offset: $scrollOffset',
      ),
      onMMBScrollStateChanged: (scrolling) => debugPrint(
        'Is scrolling: $scrolling',
      ),
      onMMBScrollCursorPositionUpdate: (localCursorOffset, scrollActivity) => debugPrint(
            'Cursor position: $localCursorOffset\n'
            'Scroll activity: $scrollActivity',
      ),
      enableMMBScrolling: true,
      enableKeyboardScrolling: true,
      enableCustomMouseWheelScrolling: true,
      mmbScrollConfig: MMBScrollConfig(
        customScrollCursor: useSystemCursor ? null : const DefaultCustomScrollCursor(),
      ),
      keyboardScrollConfig: KeyboardScrollConfig(
        arrowsScrollAmount: 250.0,
        homeScrollDurationBuilder: (currentScrollOffset, minScrollOffset) {
          return const Duration(milliseconds: 100);
        },
        endScrollDurationBuilder: (currentScrollOffset, maxScrollOffset) {
          return const Duration(milliseconds: 2000);
        },
      ),
      customMouseWheelScrollConfig: const CustomMouseWheelScrollConfig(
        scrollAmountMultiplier: 2.0,
      ),
      child: ScrollConfiguration(
        behavior: const CustomScrollBehaviour(),
        child: GridView(
          controller: controller,
          physics: const NeverScrollableScrollPhysics(),
          scrollDirection: axis,
          padding: const EdgeInsets.all(24.0),
          gridDelegate: const SliverGridDelegateWithMaxCrossAxisExtent(
            maxCrossAxisExtent: 400.0,
            mainAxisExtent: 400.0,
          ),
          children: buildScrollableItemList(axis),
        ),
      ),
    );
```

<br>

### Requirements

- The `ImprovedScrolling` Widget must be added as a parent of your scrollable Widget (`List/Grid/SingleChildScrollView/etc`).
- You must create and provide the same scroll controller to both the `ImprovedScrolling` Widget and your scrollable Widget.

* _Optional_: If you want to programatically scroll when rotating the mouse wheel and not let the framework manage the scrolling, you can set `physics: NeverScrollableScrollPhysics()` to your scrollable and then set `enableCustomMouseWheelScrolling: true` on `ImprovedScrolling` to enable this feature.

<br>

### Features:

- #### Scrolling using the keyboard (Arrows, Page{Up, Down}, Spacebar, Home, End)

  > You need to set `enableKeyboardScrolling: true` and then you can configure the scrolling amount, duration and curve by using `keyboardScrollConfig: KeyboardScrollConfig(...)`

- #### Scrolling using the middle mouse button ("auto-scrolling")

  > You need to set `enableMMBScrolling: true` and then you can configure the scrolling delay, velocity, acceleration and cursor appearance and size by using `mmbScrollConfig: MMBScrollConfig(...)`  
  > <br>
  >
  > There is also a `DefaultCustomScrollCursor` class which is a default custom cursor implementation

- #### Programatically scroll using the mouse wheel

  > You need to set `enableCustomMouseWheelScrolling: true` and then you can configure the scrolling speed, duration, curve and throttling time by using `customMouseWheelScrollConfig: CustomMouseWheelScrollConfig(...)`

- #### Horizontal scrolling using Left/Right arrows or Shift + mouse wheel

  > Requires `enableKeyboardScrolling: true` and `enableCustomMouseWheelScrolling: true` to be set.

<br>

### Callbacks:

#### Other than the above features, there are also a few callbacks available on the `ImprovedScrolling` Widget:

<br>

```dart
// Triggers whenever the ScrollController scrolls, no matter how or why
onScroll: (double scrollOffset) => debugPrint(
  'Scroll offset: $scrollOffset',
),

// Triggers whenever the middle mouse button scrolling feature is activated or deactivated
onMMBScrollStateChanged: (bool scrolling) => debugPrint(
  'Is scrolling: $scrolling',
),

// Triggers whenever the cursor is moved on the scrollable area, while the
// middle mouse button feature is active and is scrolling
//
// We also get the current scroll activity (idle or moving up/down/left/right)
// at the time the cursor moves
onMMBScrollCursorPositionUpdate: (
  Offset localCursorOffset,
  MMBScrollCursorActivity scrollActivity,
) => debugPrint(
    'Cursor position: $localCursorOffset\n'
    'Scroll activity: $scrollActivity',
 ),
```

## _**License**_

[MIT](https://github.com/adrianflutur/flutter_improved_scrolling/blob/main/LICENSE)


CHANGELOG.md
## 0.0.3

- Added one throttler for each mouse wheel scrolling direction for being able to throttle independently

## 0.0.2

- Added throttling for mouse wheel scroll events when using custom mouse wheel scrolling as an attempt to make the wheel scrolling smooth

- Specify in the example app that it is also possible to use HOME and END keys for keyboard scrolling

## 0.0.1

- Initial release.


lib/src/config.dart
import 'package:flutter/animation.dart';
import 'package:flutter/material.dart';

import 'custom_scroll_cursor.dart';

/// Middle mouse button scrolling configuration
class MMBScrollConfig {
  /// Default values
  const MMBScrollConfig({
    this.customScrollCursor,
    this.idleCursorAreaSize = 30.0,
    this.autoScrollDelay = Duration.zero,
    this.velocityBackpropagationPercent = 30.0 / 100.0,
    this.decelerationForce = 500.0,
  });

  /// Custom cursor specified by the user
  ///
  /// If null, default system cursors will be used instead
  final CustomScrollCursor? customScrollCursor;

  /// Size of the area where the cursor changes to idle
  /// This gets split in half (one half up, one half down)
  final double idleCursorAreaSize;

  /// The delay between sequential auto scrolls.
  ///
  /// Affects scrolling framerate (higher delay = more staggered/laggy scrolling)
  final Duration autoScrollDelay;

  /// Percent of how much speed to extract from the
  /// last scrolling's velocity and add to the next velocity
  ///
  /// Affects speed
  final double velocityBackpropagationPercent;

  /// Amount of friction
  ///
  /// Affects how smooth to initiate the scrolling when getting out
  /// of the idle zone, but will also affect overall speed
  ///
  /// A lower value will make the scrolling start abruptly,
  /// while a higher value will make it smooth
  final double decelerationForce;
}

/// Keyboard scrolling configuration
class KeyboardScrollConfig {
  /// Default values
  const KeyboardScrollConfig({
    this.arrowsScrollAmount = 200.0,
    this.arrowsScrollDuration = const Duration(milliseconds: 200),
    this.pageUpDownScrollAmount = 500.0,
    this.pageUpDownScrollDuration = const Duration(milliseconds: 200),
    this.spaceScrollAmount = 600.0,
    this.spaceScrollDuration = const Duration(milliseconds: 200),
    this.defaultHomeEndScrollDuration = const Duration(milliseconds: 500),
    this.homeScrollDurationBuilder,
    this.endScrollDurationBuilder,
    this.scrollCurve = Curves.easeOutCubic,
  });

  /// Amount to scroll when pressing arrow keys
  final double arrowsScrollAmount;

  /// Duration to reach the new scroll position when using arrow keys
  final Duration arrowsScrollDuration;

  /// Amount to scroll when pressing page up/down
  final double pageUpDownScrollAmount;

  /// Duration to reach the new scroll position when using page up/down
  final Duration pageUpDownScrollDuration;

  /// Amount to scroll when pressing space
  final double spaceScrollAmount;

  /// Duration to reach the new scroll position when using space
  final Duration spaceScrollDuration;

  /// Compute duration to reach the start of the scroll view
  /// based on where the scroll offset is right now
  final Duration Function(
    double currentScrollOffset,
    double minScrollOffset,
  )? homeScrollDurationBuilder;

  /// Compute duration to reach the end of the scroll view
  /// based on where the scroll offset is right now
  final Duration Function(
    double currentScrollOffset,
    double maxScrollOffset,
  )? endScrollDurationBuilder;

  /// Default duration for home and end scrolling
  final Duration defaultHomeEndScrollDuration;

  /// Scroll curve
  final Curve scrollCurve;
}

/// Custom mouse wheel scrolling configuration
class CustomMouseWheelScrollConfig {
  /// Default values
  const CustomMouseWheelScrollConfig({
    this.scrollAmountMultiplier = 3.0,
    this.scrollDuration = const Duration(milliseconds: 400),
    this.scrollCurve = Curves.linearToEaseOut,
    this.mouseWheelTurnsThrottleTimeMs = 80,
  });

  /// Extra amount to scroll when scrolling using mouse wheel
  ///
  /// Can be negative
  final double scrollAmountMultiplier;

  /// Scroll duration
  final Duration scrollDuration;

  /// Scroll curve
  final Curve scrollCurve;

  /// Scrolling throttle duration
  final int mouseWheelTurnsThrottleTimeMs;
}


lib/src/custom_scroll_cursor.dart
import 'dart:math';

import 'package:flutter/material.dart';

/// Custom cursor widget interface
abstract class CustomScrollCursor {
  /// Size
  double get size;

  /// The widget to show when in `CursorScrollActivity.idle` state
  Widget get idle;

  /// The widget to show when in `CursorScrollActivity.scrollingUp` state
  Widget get scrollingUp;

  /// The widget to show when in `CursorScrollActivity.scrollingDown` state
  Widget get scrollingDown;

  /// The widget to show when in `CursorScrollActivity.scrollingLeft` state
  Widget get scrollingLeft;

  /// The widget to show when in `CursorScrollActivity.scrollingRight` state
  Widget get scrollingRight;
}

/// Default custom cursor widget implementation
class DefaultCustomScrollCursor implements CustomScrollCursor {
  /// Constructor
  const DefaultCustomScrollCursor({
    Color cursorColor = Colors.black,
    Color borderColor = Colors.black,
    Color backgroundColor = Colors.white,
  })  : _cursorColor = cursorColor,
        _borderColor = borderColor,
        _backgroundColor = backgroundColor;

  final Color _cursorColor;
  final Color _borderColor;
  final Color _backgroundColor;

  Widget _buildCursor({
    bool idle = false,
    required Widget icon,
  }) {
    return Container(
      decoration: idle
          ? BoxDecoration(
              color: _backgroundColor,
              shape: BoxShape.circle,
              border: Border.all(
                color: _borderColor,
              ),
            )
          : null,
      child: icon,
    );
  }

  /// Size
  @override
  double get size => 28.0;

  /// The widget to show when in `CursorScrollActivity.idle` state
  @override
  Widget get idle => Transform.rotate(
        angle: 3 * pi / 4,
        child: _buildCursor(
          idle: true,
          icon: Icon(
            Icons.zoom_out_map,
            color: _cursorColor,
            size: size - 6.0,
          ),
        ),
      );

  /// The widget to show when in `CursorScrollActivity.scrollingUp` state
  @override
  Widget get scrollingUp => _buildCursor(
        icon: Icon(
          Icons.arrow_drop_up,
          color: _cursorColor,
          size: size,
        ),
      );

  /// The widget to show when in `CursorScrollActivity.scrollingDown` state
  @override
  Widget get scrollingDown => Transform.rotate(
        angle: pi,
        child: scrollingUp,
      );

  /// The widget to show when in `CursorScrollActivity.scrollingLeft` state
  @override
  Widget get scrollingLeft => Transform.rotate(
        angle: -pi / 2,
        child: scrollingUp,
      );

  /// The widget to show when in `CursorScrollActivity.scrollingRight` state
  @override
  Widget get scrollingRight => Transform.rotate(
        angle: pi / 2,
        child: scrollingUp,
      );
}


example/lib/scrollable_page.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_improved_scrolling/flutter_improved_scrolling.dart';

class CustomScrollBehaviour extends MaterialScrollBehavior {
  const CustomScrollBehaviour();

  @override
  Widget buildScrollbar(
    BuildContext context,
    Widget child,
    ScrollableDetails details,
  ) {
    switch (getPlatform(context)) {
      case TargetPlatform.linux:
      case TargetPlatform.macOS:
        return Scrollbar(
          controller: details.controller,
          isAlwaysShown: true,
          child: child,
        );
      case TargetPlatform.windows:
        return Scrollbar(
          controller: details.controller,
          isAlwaysShown: true,
          radius: Radius.zero,
          thickness: 16.0,
          hoverThickness: 16.0,
          showTrackOnHover: true,
          child: child,
        );
      case TargetPlatform.android:
      case TargetPlatform.fuchsia:
      case TargetPlatform.iOS:
        return child;
    }
  }
}

class ScrollablePage extends StatefulWidget {
  const ScrollablePage({Key? key}) : super(key: key);

  @override
  _ScrollablePageState createState() => _ScrollablePageState();
}

class _ScrollablePageState extends State<ScrollablePage> {
  final scrollControllerVertical = ScrollController();
  final scrollControllerVertical2 = ScrollController();
  final scrollControllerHorizontal = ScrollController();
  final scrollControllerHorizontal2 = ScrollController();

  bool get isDesktop =>
      Theme.of(context).platform == TargetPlatform.linux ||
      Theme.of(context).platform == TargetPlatform.macOS ||
      Theme.of(context).platform == TargetPlatform.windows;

  TextStyle? get textStyleBig => Theme.of(context).textTheme.headline5;
  TextStyle? get textStyleMedium => Theme.of(context).textTheme.headline6;
  TextStyle? get textStyleSmall => Theme.of(context).textTheme.bodyText2;

  Axis axis = Axis.vertical;
  bool useSystemCursor = false;

  void toggleAxis() {
    setState(() {
      if (axis == Axis.vertical) {
        axis = Axis.horizontal;
      } else {
        axis = Axis.vertical;
      }
    });
  }

  void toggleCursor() {
    setState(() {
      useSystemCursor = !useSystemCursor;
    });
  }

  @override
  void dispose() {
    scrollControllerVertical.dispose();
    scrollControllerVertical2.dispose();
    scrollControllerHorizontal.dispose();
    scrollControllerHorizontal2.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          buildHeader(),
          const Divider(
            thickness: 5.0,
            color: Colors.grey,
          ),
          if (axis == Axis.vertical)
            Expanded(
              child: Row(
                children: [
                  Expanded(
                    child: buildScrollingView(
                      Axis.vertical,
                      scrollControllerVertical,
                    ),
                  ),
                  const VerticalDivider(),
                  Expanded(
                    child: buildScrollingView(
                      Axis.vertical,
                      scrollControllerVertical2,
                    ),
                  ),
                ],
              ),
            )
          else
            Expanded(
              child: Column(
                children: [
                  Expanded(
                    child: buildScrollingView(
                      Axis.horizontal,
                      scrollControllerHorizontal,
                    ),
                  ),
                  const Divider(),
                  Expanded(
                    child: buildScrollingView(
                      Axis.horizontal,
                      scrollControllerHorizontal2,
                    ),
                  ),
                ],
              ),
            ),
        ],
      ),
      floatingActionButton: Padding(
        padding: const EdgeInsets.all(10.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.end,
          children: [
            FloatingActionButton.extended(
              onPressed: toggleCursor,
              label: const Text('Toggle cursor'),
              backgroundColor: Colors.black,
              heroTag: 'FirstFAB',
            ),
            const SizedBox(height: 10.0),
            FloatingActionButton.extended(
              onPressed: toggleAxis,
              label: const Text('Toggle axis'),
              backgroundColor: Colors.black,
              heroTag: 'SecondFAB',
            ),
          ],
        ),
      ),
    );
  }

  Widget buildHeader() {
    return Card(
      margin: const EdgeInsets.all(12.0),
      child: Padding(
        padding: const EdgeInsets.all(12.0),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    'Advanced scrolling for Flutter web and desktop',
                    style: textStyleBig,
                  ),
                  Text(
                    '  1. Scroll using middle mouse button:',
                    style: textStyleMedium,
                  ),
                  Text(
                    '          - Tap MMB --> Drag --> Tap MMB, or',
                    style: textStyleSmall,
                  ),
                  Text(
                    '          - Press and hold MMB --> Drag --> Release MMB',
                    style: textStyleSmall,
                  ),
                  Text(
                    '          - Can use default system cursors,'
                    ' or custom Widget cursor',
                    style: textStyleSmall,
                  ),
                  Text(
                    '  2. Scroll using keyboard (the scrollable'
                    ' widget must have focus):',
                    style: textStyleMedium,
                  ),
                  Text(
                    '          - Arrows (up, down, left, right)',
                    style: textStyleSmall,
                  ),
                  Text(
                    '          - Space, Shift + Space (reverse)',
                    style: textStyleSmall,
                  ),
                  Text(
                    '          - PageUp, PageDown, End, Home',
                    style: textStyleSmall,
                  ),
                ],
              ),
            ),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    '3. Can also programatically scroll using mouse wheel\n'
                    '(when using NeverScrollableScrollPhysics)',
                    style: textStyleMedium,
                  ),
                  Text(
                    '4. Supports both vertical and horizontal scrolling,'
                    ' but not at the same time',
                    style: textStyleMedium,
                  ),
                  Text(
                    '5. Check out the console for events log',
                    style: textStyleMedium,
                  ),
                  const SizedBox(height: 20.0),
                  Text(
                    'Current cursor type: '
                    '${useSystemCursor ? 'system' : 'custom'}',
                    style: textStyleSmall,
                  ),
                  Text(
                    'Current scrollables orientation: ${describeEnum(axis)}',
                    style: textStyleSmall,
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget buildScrollingView(Axis axis, ScrollController controller) {
    return ImprovedScrolling(
      scrollController: controller,
      onScroll: (scrollOffset) => debugPrint(
        'Scroll offset: $scrollOffset',
      ),
      onMMBScrollStateChanged: (scrolling) => debugPrint(
        'Is scrolling: $scrolling',
      ),
      onMMBScrollCursorPositionUpdate: (localCursorOffset, scrollActivity) =>
          debugPrint(
        'Cursor position: $localCursorOffset\n'
        'Scroll activity: $scrollActivity',
      ),
      enableMMBScrolling: true,
      enableKeyboardScrolling: true,
      enableCustomMouseWheelScrolling: true,
      mmbScrollConfig: MMBScrollConfig(
        customScrollCursor:
            useSystemCursor ? null : const DefaultCustomScrollCursor(),
      ),
      keyboardScrollConfig: KeyboardScrollConfig(
        homeScrollDurationBuilder: (currentScrollOffset, minScrollOffset) {
          return const Duration(milliseconds: 100);
        },
        endScrollDurationBuilder: (currentScrollOffset, maxScrollOffset) {
          return const Duration(milliseconds: 2000);
        },
      ),
      customMouseWheelScrollConfig: const CustomMouseWheelScrollConfig(
        scrollAmountMultiplier: 4.0,
        scrollDuration: Duration(milliseconds: 350),
      ),
      child: ScrollConfiguration(
        behavior: const CustomScrollBehaviour(),
        child: GridView(
          controller: controller,
          physics: const NeverScrollableScrollPhysics(),
          scrollDirection: axis,
          padding: const EdgeInsets.all(24.0),
          gridDelegate: const SliverGridDelegateWithMaxCrossAxisExtent(
            maxCrossAxisExtent: 400.0,
            mainAxisExtent: 400.0,
          ),
          children: buildScrollableItemList(axis),
        ),
      ),
    );
  }

  List<Widget> buildScrollableItemList(Axis axis) {
    final isVertical = axis == Axis.vertical;
    final Size itemsSize;
    if (isVertical) {
      itemsSize = const Size(700.0, 300.0);
    } else {
      itemsSize = const Size(300.0, 500.0);
    }

    return [
      for (var i = 1; i <= 100; i++)
        buildScrollableItem(
          size: itemsSize,
          child: Center(
            child: Text(
              '$i',
              style: textStyleMedium,
            ),
          ),
        ),
    ];
  }

  Widget buildScrollableItem({
    required Size size,
    required Widget child,
  }) {
    return GridTile(
      child: Container(
        margin: const EdgeInsets.only(
          top: 24.0,
          left: 24.0,
          right: 24.0,
        ),
        padding: const EdgeInsets.all(12.0),
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(10.0),
          border: Border.all(color: Colors.black.withAlpha(10), width: 2.0),
          color: Colors.white,
        ),
        width: size.width,
        height: size.height,
        child: child,
      ),
    );
  }
}
